File: FreshVienna/FTUIMainViewController.m
//
//  FTUIMainViewController.m
//  FreshVienna
//
//  Created by Thibaut on 5/11/14.
//  Copyright (c) 2014 Thibaut. All rights reserved.
//

#import "FTUIMainViewController.h"
#import "FTUrlLoader.h"
#import "FTUIChannelDelegate.h"
#import <WebKit/WebKit.h>
#import "FTOPMLReader.h"
#import "FTOPMLItem.h"
#import "ImageAndTextCell.h"
#import "FTUIOutlineViewDataSource.h"
#import "FTUITreeItem.h"

@interface FTUIMainViewController ()
{
    // delegates
    FTUIChannelDelegate *channelDelegate;

    // data
    NSMutableArray *loaders;
    FTUIOutlineViewDataSource *outlineDataSource;
    
    
    //views
    NSSplitView *splitView;
    NSTableView *tableView;
    NSScrollView *tableContainer;
    NSScrollView *webContainer;
    NSOutlineView *outlineView;
    
    // temp views
    NSScrollView *newTableContainer;
    
    
    // WebView HTML and CSS
    NSString *htmlTemplate;
    NSString *cssTemplate;
    
    NSMutableArray *treeArray;
    
}

@property WebView *webView;
@property NSArray *urls;
@end

@implementation FTUIMainViewController
@synthesize urls, webView;

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        loaders = [NSMutableArray array];
    }
    
    
    FTOPMLReader *reader = [[FTOPMLReader alloc] init];
    NSBundle *mainBundle = [NSBundle mainBundle];
    NSString *filePath = [mainBundle pathForResource:@"Test" ofType:@"opml"];

    
    self.urls = [reader loadUrl:filePath];
    
    treeArray = [NSMutableArray array];
    
    for (FTOPMLItem* url in self.urls) {
        FTUrlLoader *urlLoader = [[FTUrlLoader alloc] initWithUrl:url.xmlUrl];
        [loaders addObject:urlLoader];
        [treeArray addObject:[FTUITreeItem treeItemFromItem:url loader:urlLoader]];
    }
    
    
    
    NSError *error;
    htmlTemplate = [NSString stringWithContentsOfFile:[mainBundle pathForResource:@"template" ofType:@"html"] encoding:NSUTF8StringEncoding error:&error];
    
    cssTemplate = [NSString stringWithContentsOfFile:[mainBundle pathForResource:@"stylesheet" ofType:@"css"] encoding:NSUTF8StringEncoding error:&error];

    
    return self;
}

-(void)loadView
{
    
    CGRect frame = [NSScreen mainScreen].frame;
    self.view = [[NSView alloc] initWithFrame:frame];
    
    CGRect outlineFrame;
    CGRect remainingFrame;
    CGRectDivide(frame, &outlineFrame, &remainingFrame, 150, CGRectMinXEdge);
    
    
    NSTableColumn *column =[[NSTableColumn alloc]initWithIdentifier:@"1"];
    
    
    // outline view
    newTableContainer = [[NSScrollView alloc] initWithFrame:outlineFrame];
    [newTableContainer setAutoresizingMask: NSViewHeightSizable];
    
    
    
    
    
    outlineFrame.origin.y = - outlineFrame.origin.y;
    outlineView = [[NSOutlineView alloc] initWithFrame:outlineFrame];
    
    
    outlineDataSource = [[FTUIOutlineViewDataSource alloc] init];
    outlineDataSource.data = treeArray;
    
    [outlineView setDataSource:(id<NSOutlineViewDataSource>)outlineDataSource];
    [outlineView setDelegate:(id<NSOutlineViewDelegate>)self];
    
    // set the first column's cells as `ImageAndTextCell`s
    ImageAndTextCell* iatc = [[ImageAndTextCell alloc] init];
    [iatc setEditable:NO];
    
    [newTableContainer setDocumentView:outlineView];
    
    
    [outlineView addTableColumn:column];
    [outlineView setIntercellSpacing:NSMakeSize(0, 0)];
    [[[outlineView tableColumns] objectAtIndex:0] setIdentifier:@"name"];
    [[[outlineView tableColumns] objectAtIndex:0] setDataCell:iatc];
    
    
//    [self.view addSubview:newTableContainer];

    
    
    
    
    
    
    
    
    
    
    
    // table view
    
    
    tableContainer = [[NSScrollView alloc] init];
    [tableContainer setAutoresizingMask:NSViewWidthSizable | NSViewHeightSizable];
    
    tableView = [[NSTableView alloc] initWithFrame:tableContainer.frame];
    [tableView setAutoresizingMask:NSViewWidthSizable | NSViewHeightSizable];
    
    [tableView setAllowsEmptySelection:NO];
    [tableView setAllowsMultipleSelection:NO];
    [tableView addTableColumn:column];
    [tableView setIntercellSpacing:NSSizeFromCGSize(CGSizeMake(1.0, 20.0))];
    [tableView setSelectionHighlightStyle:NSTableViewSelectionHighlightStyleNone];
                                    
                                    
    channelDelegate = [[FTUIChannelDelegate alloc] init];
    
    
    [tableContainer setDocumentView:tableView];
    tableView.backgroundColor = [NSColor colorWithCalibratedRed:173/255.0 green:178/255.0 blue:187/255.0 alpha:1];
    
    [tableView setTarget:self];
    [tableView setDoubleAction:@selector(doubleClickOnRssItem:)];
    
    
    
    
    
    // web view
    
    
    webContainer = [[NSScrollView alloc] init];
    [webContainer setAutoresizingMask:NSViewWidthSizable | NSViewHeightSizable];
    
    self.webView = [[WebView alloc] initWithFrame:webContainer.frame];
    [self.webView setAutoresizingMask:NSViewWidthSizable | NSViewHeightSizable];

    
    [self.webView setPolicyDelegate:self];
    
    [webContainer setDocumentView:webView];
    [tableView reloadData];
    
    
    
    splitView = [[NSSplitView alloc] initWithFrame:frame];
    [splitView setDelegate:self];
    [splitView setVertical:YES];
    [splitView setAutoresizingMask:NSViewWidthSizable | NSViewHeightSizable];
    [splitView addSubview:newTableContainer];
    [splitView addSubview:tableContainer];
    [splitView addSubview:webContainer];
    [splitView setDividerStyle:NSSplitViewDividerStyleThin];
    [splitView adjustSubviews];
    [splitView setPosition:outlineFrame.size.width ofDividerAtIndex:0];
    [splitView setPosition:(remainingFrame.origin.x + remainingFrame.size.width/2) ofDividerAtIndex:1];

    [self.view addSubview:splitView];
    

    
    

    
    
    
    
    NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
    [center addObserverForName:@"rssItemClicked"
                        object:nil
                         queue:nil
                    usingBlock:^(NSNotification *notification)
     {
         
         
         FTRSSItemAttributes* item = notification.object;
         [self loadWebView:item];
     }];

    self.webView.frameLoadDelegate = self;
    
    [center addObserverForName:@"loading updated" object:nil queue:nil usingBlock:^(NSNotification *notification)
    {
        [outlineView reloadData];
    }];
}






/*******************************************************
 *
 * OUTLINE-VIEW DELEGATE
 *
 *******************************************************/

- (BOOL)outlineView:(NSOutlineView *)outlineView shouldSelectItem:(id)item
{
    return YES;
}

- (BOOL)outlineView:(NSOutlineView *)outlineView isGroupItem:(id)item
{
    return NO;
}

- (void)outlineViewSelectionDidChange:(NSNotification *)notification
{
    NSInteger selected = [outlineView selectedRow];
    if (selected > -1)
    {
        FTUrlLoader *loader = [loaders objectAtIndex:selected];
        channelDelegate.rssItems = loader.rssItems;
        channelDelegate.feedItem = loader.channelAttributes;
        
        [tableView scrollToBeginningOfDocument:tableView];
        
        if (tableView.dataSource != channelDelegate)
        {
            [tableView setDataSource:channelDelegate];
            [tableView setDelegate:channelDelegate];
        }
        [tableView reloadData];
        
    }

}

- (void)outlineView:(NSOutlineView *)outlineView willDisplayCell:(id)cell forTableColumn:(NSTableColumn *)tableColumn item:(id)item {
    [cell setDrawsBackground:YES];
    [cell setTextColor:[NSColor blackColor]];
    
}

































#pragma SplitView Delegate

- (CGFloat)splitView:(NSSplitView *)splitView constrainMinCoordinate:(CGFloat)proposedMinimumPosition ofSubviewAt:(NSInteger)dividerIndex;
{
    if (dividerIndex == 0)
    {
        return proposedMinimumPosition + 50;
    }
    return proposedMinimumPosition + 200;
}
/*
 * Controls the minimum size of the right subview (or lower subview in a horizonal NSSplitView)
 */
- (CGFloat)splitView:(NSSplitView *)splitView constrainMaxCoordinate:(CGFloat)proposedMaximumPosition ofSubviewAt:(NSInteger)dividerIndex;
{
    return proposedMaximumPosition - 100;
}


#pragma TableDataSource Delegate
- (NSInteger)numberOfRowsInTableView:(NSTableView *)tableView {
    return self.urls.count;
}


#pragma TableView Delegate
// cell based
- (id)tableView:(NSTableView *)aTableView objectValueForTableColumn:(NSTableColumn *)aTableColumn row:(NSInteger)rowIndex
{
    return ((FTOPMLItem*)[self.urls objectAtIndex:rowIndex]).title;
}


- (void)loadWebView:(FTRSSItemAttributes*)item
{
    
    NSMutableString * htmlText = [[NSMutableString alloc] initWithString:@"<!DOCTYPE html><html><head><meta charset=\"UTF-8\" />"];
	if (cssTemplate != nil)
	{
		[htmlText appendString:@"<style type=\"text/css\">"];
		[htmlText appendString:cssTemplate];
		[htmlText appendString:@"</style>"];
	}

	[htmlText appendString:@"<meta http-equiv=\"Pragma\" content=\"no-cache\">"];
	[htmlText appendString:@"</head><body>"];
		
    // Load the selected HTML template for the current view style and plug in the current
    // article values and style sheet setting.
    NSString * htmlArticle;
    htmlArticle = [htmlTemplate stringByReplacingOccurrencesOfString:@"$ArticleLink$" withString:[item link]];
    htmlArticle = [htmlArticle stringByReplacingOccurrencesOfString:@"$ArticleTitle$" withString:[item title]];
    htmlArticle = [htmlArticle stringByReplacingOccurrencesOfString:@"$ArticleDate$" withString:[item pubDate]];
    htmlArticle = [htmlArticle stringByReplacingOccurrencesOfString:@"$ArticleBody$" withString:[item description]];
    
    
    NSInteger selected = [outlineView selectedRow];
    if (selected > -1)
    {
        FTUrlLoader *loader = [loaders objectAtIndex:selected];
        htmlArticle = [htmlArticle stringByReplacingOccurrencesOfString:@"$FeedTitle$" withString:[loader.channelAttributes title]];
    }
    
    [htmlText appendString:htmlArticle];
	[htmlText appendString:@"</body></html>"];
    
    [self.webView.mainFrame loadHTMLString:htmlText baseURL:[NSURL URLWithString:item.link]];
}


- (void)webView:(WebView *)sender didFinishLoadForFrame:(WebFrame *)frame
{
    //get the scroll view that contains the frame contents
    NSScrollView* scrollView = [[[[webView mainFrame] frameView] documentView] enclosingScrollView];
    [[scrollView verticalScroller] setControlSize: NSSmallControlSize];
    [[scrollView horizontalScroller] setControlSize: NSSmallControlSize];
}

- (void)webView:(WebView *)sender decidePolicyForNavigationAction:(NSDictionary *)actionInformation request:(NSURLRequest *)request frame:(WebFrame *)frame decisionListener:(id <WebPolicyDecisionListener>)listener
{
    if ([[actionInformation objectForKey:WebActionNavigationTypeKey] intValue]  == WebNavigationTypeLinkClicked) {
        [listener ignore];
        [[NSWorkspace sharedWorkspace] openURL:[request URL]];
    }
    else
        [listener use];
}




-(void)doubleClickOnRssItem:(id)object
{
    FTRSSItemAttributes* item = [channelDelegate.rssItems objectAtIndex:[tableView clickedRow]];
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:item.link]];

}

@end

File: FreshSwiftVienna/FTUIMainViewController.swift
//
//  FTUIMainViewController.swift
//  FreshVienna
//
//  Created by Thibaut on 6/2/14.
//  Copyright (c) 2014 Thibaut. All rights reserved.
//

import Foundation
import Cocoa
import WebKit

class FTUIMainViewController : NSViewController, NSTableViewDataSource, NSTableViewDelegate, NSOutlineViewDelegate
{
    
    // delegates
    var channelDelegate:FTUIChannelDelegate;
    var webViewDelegate:FTUIWebViewDelegate;
    var splitViewDelegate:FTUISplitViewDelegate;
    
    // data
    var urls:FTOPMLItem[];
    var loaders:FTURLLoader[];
    var outlineDataSource:FTUIOutlineViewDataSource;
    var treeArray:FTUITreeItem[];
    let queue:NSOperationQueue;
    
    //views
    var splitView:NSSplitView;
    var tableView:NSTableView;
    var tableContainer:NSScrollView;
    var webView:WebView;
    var webContainer:NSScrollView;
    var outlineView:NSOutlineView;
    var outlineContainer:NSScrollView;
    
    // WebView HTML and CSS
    var htmlTemplate:String?;
    var cssTemplate:String?;
    
    init(nibName nibNameOrNil: String!, bundle nibBundleOrNil: NSBundle!)
    {
        splitView = NSSplitView();
        tableView = NSTableView();
        tableContainer = NSScrollView();
        webContainer = NSScrollView();
        outlineView = NSOutlineView();
        outlineContainer = NSScrollView();
        webView = WebView();
        treeArray = [];
        loaders = [];
        urls = [];
        queue = NSOperationQueue();
        channelDelegate = FTUIChannelDelegate();
        outlineDataSource = FTUIOutlineViewDataSource();
        webViewDelegate = FTUIWebViewDelegate();
        splitViewDelegate = FTUISplitViewDelegate();
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil);
        loadRessources();
    }
    
    
    func loadRessources()
    {
        var reader:FTOPMLReader = FTOPMLReader();
        var error:NSError?;
        let mainBundle = NSBundle.mainBundle();
        
        if let filePath:String = mainBundle.pathForResource("Test", ofType: "opml")
        {
            urls = reader.loadUrl(filePath);
        }
        
//        urls = [FTOPMLItem(xmlUrl:"https://news.ycombinator.com/rss", title:"Hacker news")];
        
        
        if let htmlTemplatePath:String = mainBundle.pathForResource("template", ofType: "html")
        {
            htmlTemplate = String.stringWithContentsOfFile(htmlTemplatePath, encoding: NSUTF8StringEncoding, error: &error)
        } else
        {
            htmlTemplate = "";
        }
        if let cssTemplatePath:String = mainBundle.pathForResource("stylesheet", ofType: "css")
        {
            cssTemplate = String.stringWithContentsOfFile(cssTemplatePath, encoding: NSUTF8StringEncoding, error: &error)
        } else
        {
            cssTemplate = "";
        }


    }
    
    override func loadView()
    {
        
        println("loadView");
        
        var frame = NSScreen.mainScreen().frame;
        self.view = NSView(frame:frame);
        
        
        var outlineFrame:CGRect = CGRect();
        var remainingFrame:CGRect = CGRect();
        CGRectDivide(frame, &outlineFrame, &remainingFrame, 240, CGRectEdge.MinXEdge);
        
        let column:NSTableColumn = NSTableColumn(identifier:"1");

        // outline view
        setupOutlineView(&outlineFrame);
        outlineView.setDelegate(self);
        
        // set the first column's cells as `ImageAndTextCell`s
        var iatc:FTImageAndTextCell = FTImageAndTextCell();
        iatc.editable = false;

        outlineContainer.documentView = outlineView;

        outlineView.addTableColumn(column);
        outlineView.intercellSpacing = NSSizeFromCGSize(CGSizeMake(0, 0));
        (outlineView.tableColumns[0] as NSTableColumn).identifier = "name";
        (outlineView.tableColumns[0] as NSTableColumn).dataCell = iatc;
        outlineView.headerView = nil;
        
        // table view
        
        tableContainer.autoresizingMask = NSAutoresizingMaskOptions.ViewWidthSizable | NSAutoresizingMaskOptions.ViewHeightSizable;

        tableView.frame = tableContainer.frame;
        tableView.autoresizingMask = NSAutoresizingMaskOptions.ViewWidthSizable | NSAutoresizingMaskOptions.ViewHeightSizable;

        tableView.allowsEmptySelection = false;
        tableView.allowsMultipleSelection = false;
        tableView.addTableColumn(column);
        tableView.intercellSpacing = NSSizeFromCGSize(CGSizeMake(1.0, 20.0));
        tableView.selectionHighlightStyle = NSTableViewSelectionHighlightStyle.None;
        

        tableContainer.documentView = tableView;
        tableView.backgroundColor = NSColor(calibratedRed: 173/255.0, green: 178/255.0, blue: 187/255.0, alpha: 1);
            
        tableView.target = self;
        tableView.doubleAction = Selector("doubleClickOnRssItem");
        
        
        // web view
        
        webContainer.autoresizingMask = NSAutoresizingMaskOptions.ViewWidthSizable | NSAutoresizingMaskOptions.ViewHeightSizable;
        webView.frame = webContainer.frame;
        webView.autoresizingMask = NSAutoresizingMaskOptions.ViewWidthSizable | NSAutoresizingMaskOptions.ViewHeightSizable;
        webView.policyDelegate = webViewDelegate;
        webContainer.documentView = webView;

        
        tableView.reloadData();
        
        setupSplitView(frame);
        splitView.setPosition(outlineFrame.size.width, ofDividerAtIndex: 0);
        splitView.setPosition((remainingFrame.origin.x + remainingFrame.size.width/2), ofDividerAtIndex:1);

        
        self.view.addSubview(splitView);
        
        queue.maxConcurrentOperationCount = 15;
        
        for url in urls {
            var urlLoader:FTURLLoader = FTURLLoader(urlPath:url.xmlUrl);
            loaders.append(urlLoader);
            var asyncLoadOperation = FTURLOperation(loader: urlLoader);
            queue.addOperation(asyncLoadOperation);
            treeArray.append(FTUITreeItem(loader: urlLoader, item: url));
        }
        
        outlineDataSource.data = treeArray;
        outlineView.setDataSource(outlineDataSource);
        
        
        let center:NSNotificationCenter = NSNotificationCenter.defaultCenter();
        
        
        let completionBlock: (NSNotification!) -> Void = { notification in
            var item:FTRSSItem = notification.object as FTRSSItem;
            self.loadWebView(item);
        };
        
        center.addObserverForName("rssItemClicked", object: nil, queue: nil, usingBlock: completionBlock);
        
        self.webView.frameLoadDelegate = self;
        self.outlineView.reloadData();

        var context:KVOContext = KVOContext();
        queue.addObserver(self, forKeyPath: "operationCount", options: NSKeyValueObservingOptions.New, kvoContext: context);

    }
    
    
    func setupSplitView(frame:CGRect)
    {
        splitView.frame = frame;
        splitView.delegate = splitViewDelegate;
        splitView.vertical = true;
        splitView.autoresizingMask = NSAutoresizingMaskOptions.ViewWidthSizable | NSAutoresizingMaskOptions.ViewHeightSizable;
        
        splitView.addSubview(outlineContainer);
        splitView.addSubview(tableContainer);
        splitView.addSubview(webContainer);
        splitView.dividerStyle = NSSplitViewDividerStyle.Thin;
        splitView.adjustSubviews();

    }
    
    
    func setupOutlineView(inout outlineFrame:CGRect)
    {
        outlineContainer.frame = outlineFrame;
        outlineContainer.autoresizingMask = NSAutoresizingMaskOptions.ViewHeightSizable;
        
        outlineFrame.origin.y = -outlineFrame.origin.y;
        outlineView.frame = outlineFrame;

    }
    
    
    override func observeValueForKeyPath(keyPath: String!, ofObject object: AnyObject!, change: NSDictionary!, context: CMutableVoidPointer)
    {
        let queueObject = object as NSOperationQueue;
        
        if queueObject == self.queue && keyPath == "operationCount"
        {
            if self.queue.operationCount == 0
            {
                println("queue has completed");
                self.outlineView.reloadData();
            } else
            {
                println("queue not done");
            }
        } else
        {
            super.observeValueForKeyPath(keyPath, ofObject: object, change: change, context: context);
        }
    }
    
    

    
    
    /*******************************************************
    *
    * OUTLINE-VIEW DELEGATE
    *
    *******************************************************/
    
    func outlineView(outlineView: NSOutlineView!, heightOfRowByItem item: AnyObject!) -> CGFloat
    {
        return 23;
    }
    
    func outlineView(outlineView: NSOutlineView!, shouldSelectItem item: AnyObject!) -> Bool
    {
        return true;
    }
    
    func outlineView(outlineView: NSOutlineView!, isGroupItem item: AnyObject!) -> Bool
    {
        return false;
    }

    func outlineViewSelectionDidChange(notification: NSNotification!)
    {
        var selected = self.outlineView.selectedRow;
        if (selected > -1)
        {
            var loader = loaders[selected];
            channelDelegate.rssItems = loader.rssItems;
            channelDelegate.feedItem = loader.rssFeed;
            
            tableView.scrollToBeginningOfDocument(tableView);

            tableView.setDataSource(channelDelegate);
            tableView.setDelegate(channelDelegate)
            
            tableView.reloadData();
        }
    }
    
    func outlineView(outlineView: NSOutlineView!, willDisplayCell cell: AnyObject!, forTableColumn tableColumn: NSTableColumn!, item: AnyObject!)
    {
        (cell as NSTextFieldCell).drawsBackground = true;
        (cell as NSTextFieldCell).textColor = NSColor.blackColor();
    }
    
        
    /*******************************************************
    *
    * TABLE VIEW DATA SOURCE
    *
    *******************************************************/

    func numberOfRowsInTableView(tableView: NSTableView!) -> Int
    {
        return urls.count;
        
    }
    
    /*******************************************************
    *
    * TABLE VIEW DELEGATE
    *
    *******************************************************/

    
    func tableView(tableView: NSTableView!, objectValueForTableColumn tableColumn: NSTableColumn!, row: Int) -> AnyObject!
    {
        return urls[row].title;
    }
    
    
    
    /*******************************************************
    *
    * WEB VIEW DELEGATE
    *
    *******************************************************/

    func loadWebView(item:FTRSSItem)
    {
        var htmlText:String = "<!DOCTYPE html><html><head><meta charset=\"UTF-8\" />";
        
        if (cssTemplate)
        {
            htmlText += "<style type=\"text/css\">";
            htmlText += cssTemplate!;
            htmlText += "</style>";
        }
        
        htmlText += "<meta http-equiv=\"Pragma\" content=\"no-cache\">";
        htmlText += "</head><body>";
        
        
        var htmlArticle = htmlTemplate!.stringByReplacingOccurrencesOfString("$ArticleLink$", withString: item.link);
        htmlArticle = htmlArticle.stringByReplacingOccurrencesOfString("$ArticleTitle$", withString: item.title);
        htmlArticle = htmlArticle.stringByReplacingOccurrencesOfString("$ArticleDate$", withString: item.pubDate);
        htmlArticle = htmlArticle.stringByReplacingOccurrencesOfString("$ArticleBody$", withString: item.description);
        
        
        var selectedRow = outlineView.selectedRow;
        if (selectedRow > -1)
        {
            var loader = loaders[selectedRow];
            htmlArticle = htmlArticle.stringByReplacingOccurrencesOfString("$FeedTitle$", withString: loader.rssFeed.title);
        }
        
        htmlText += "\(htmlArticle)";
        htmlText += "</body></html>";
        
        self.webView.mainFrame.loadHTMLString(htmlText, baseURL: NSURL.URLWithString(item.link));

    }
    
    func doubleClickOnRssItem()
    {
        var item = channelDelegate.rssItems[tableView.clickedRow];
        openUrlInBackground(item.link);
    }
    
    func openUrlInBackground(urlPath:String)
    {
        var url:NSURL = NSURL.URLWithString(urlPath);
        var sharedWorkspace:NSWorkspace = NSWorkspace.sharedWorkspace();
        var options:NSWorkspaceLaunchOptions = (NSWorkspaceLaunchOptions.WithoutActivation | NSWorkspaceLaunchOptions.Default);
        sharedWorkspace.openURLs([url], withAppBundleIdentifier: nil, options: options, additionalEventParamDescriptor: nil, launchIdentifiers: nil);
    }
    


}